From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: alichinese_admin <lixinyu44@huawei.com>
Date: Mon, 14 Sep 2026 21:00:00 +0800
Subject: [PATCH] Fix implicit declarations for clang 16+ / gcc 14+
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

clang 16+ (and gcc 14+) turn implicit-function-declaration and
implicit-int into errors in C99+ mode, and openEuler's clang toolchain
is 17.0.6, so tcp-wrappers 7.6 fails to compile on every clang target
(https://gitcode.com/openeuler/yocto-meta-openeuler/issues/1410).

The existing 0001-Fix-implicit-function-declaration-warnings.patch
only covers a few files and its safe_finger.c hunk does not include
the headers that the later tcp_wrappers-7.6-initgroups.patch calls
need. A full recipe-equivalent build with
-Werror=implicit-function-declaration -Werror=implicit-int shows the
remaining gaps; fix them all by adding the missing system headers:

- <unistd.h> in clean_exit.c/percent_x.c/update.c (sleep, getpid),
  rfc931.c (close, alarm), tcpd.c (execv), tcpdchk.c and tcpdmatch.c
  (getopt)
- <stdlib.h> in options.c (setenv), tcpdmatch.c (free), inetcf.c
  (malloc), safe_finger.c (putenv)
- <netdb.h> in hosts_access.c (innetgr, NETGROUP builds)
- <sys/wait.h> + <fcntl.h> in safe_finger.c (wait, open)
- <grp.h> in safe_finger.c (initgroups/setgroups via the spec patch)
- "scaffold.h" in inetcf.c (check_path)
- explicit int return type on main() in safe_finger.c and try-from.c
- explicit void return types on the fakelog.c syslog stubs

Verified by replicating the recipe do_configure/do_compile on the
fully patched tree (glibc, gcc with the two -Werror flags standing in
for clang 17): make all and the libwrap.so link now succeed.

Upstream-Status: Inappropriate [upstream is dead]
Signed-off-by: alichinese_admin <lixinyu44@huawei.com>

---
diff --git a/clean_exit.c b/clean_exit.c
index cb9d4f5..41caaf0 100644
--- a/clean_exit.c
+++ b/clean_exit.c
@@ -13,6 +13,7 @@ static char sccsid[] = "@(#) clean_exit.c 1.4 94/12/28 17:42:19";
 #endif
 
 #include <stdio.h>
+#include <unistd.h>
 
 extern void exit();
 
diff --git a/fakelog.c b/fakelog.c
index fa9e06e..6c5315f 100644
--- a/fakelog.c
+++ b/fakelog.c
@@ -17,7 +17,7 @@ static char sccsid[] = "@(#) fakelog.c 1.3 94/12/28 17:42:21";
 
 /* ARGSUSED */
 
-openlog(name, logopt, facility)
+void openlog(name, logopt, facility)
 char   *name;
 int     logopt;
 int     facility;
@@ -27,7 +27,7 @@ int     facility;
 
 /* vsyslog - format one record */
 
-vsyslog(severity, fmt, ap)
+void vsyslog(severity, fmt, ap)
 int     severity;
 char   *fmt;
 va_list ap;
@@ -43,6 +43,7 @@ va_list ap;
 
 /* VARARGS */
 
+void
 VARARGS(syslog, int, severity)
 {
     va_list ap;
@@ -56,7 +57,7 @@ VARARGS(syslog, int, severity)
 
 /* closelog - dummy */
 
-closelog()
+void closelog()
 {
     /* void */
 }
diff --git a/hosts_access.c b/hosts_access.c
index 1af29bf..a9a3f16 100644
--- a/hosts_access.c
+++ b/hosts_access.c
@@ -27,6 +27,7 @@ static char sccsid[] = "@(#) hosts_access.c 1.21 97/02/12 02:13:22";
 #include <sys/param.h>
 #include <netinet/in.h>
 #include <arpa/inet.h>
+#include <netdb.h>
 #include <rpcsvc/ypclnt.h>
 #include <stdio.h>
 #include <syslog.h>
diff --git a/inetcf.c b/inetcf.c
index eda8a32..29d55ce 100644
--- a/inetcf.c
+++ b/inetcf.c
@@ -14,12 +14,14 @@ static char sccsid[] = "@(#) inetcf.c 1.7 97/02/12 02:13:23";
 #include <stdio.h>
 #include <errno.h>
 #include <string.h>
+#include <stdlib.h>
 
 extern int errno;
 extern void exit();
 
 #include "tcpd.h"
 #include "inetcf.h"
+#include "scaffold.h"
 
  /*
   * Network configuration files may live in unusual places. Here are some
diff --git a/options.c b/options.c
index d8b2aa8..5053431 100644
--- a/options.c
+++ b/options.c
@@ -41,6 +41,7 @@ static char sccsid[] = "@(#) options.c 1.17 96/02/11 17:01:31";
 #include <netinet/in.h>
 #include <netdb.h>
 #include <stdio.h>
+#include <stdlib.h>
 #include <unistd.h>
 #include <syslog.h>
 #include <pwd.h>
diff --git a/percent_x.c b/percent_x.c
index c95a1ea..560d564 100644
--- a/percent_x.c
+++ b/percent_x.c
@@ -17,6 +17,7 @@ static char sccsid[] = "@(#) percent_x.c 1.4 94/12/28 17:42:37";
 /* System libraries. */
 
 #include <stdio.h>
+#include <unistd.h>
 #include <syslog.h>
 #include <string.h>
 
diff --git a/rfc931.c b/rfc931.c
index 26f57b3..5a6e515 100644
--- a/rfc931.c
+++ b/rfc931.c
@@ -16,6 +16,7 @@ static char sccsid[] = "@(#) rfc931.c 1.10 95/01/02 16:11:34";
 /* System libraries. */
 
 #include <stdio.h>
+#include <unistd.h>
 #include <syslog.h>
 #include <sys/types.h>
 #include <sys/socket.h>
diff --git a/safe_finger.c b/safe_finger.c
index f365922..052230e 100644
--- a/safe_finger.c
+++ b/safe_finger.c
@@ -22,11 +22,16 @@ static char sccsid[] = "@(#) safe_finger.c 1.4 94/12/28 17:42:41";
 
 #include <sys/types.h>
 #include <sys/stat.h>
+#include <sys/wait.h>
+#include <fcntl.h>
 #include <signal.h>
 #include <stdio.h>
 #include <ctype.h>
+#include <stdlib.h>
 #include <pwd.h>
 #include <syslog.h>
+#include <unistd.h>
+#include <grp.h>
 
 extern void exit();
 extern int pipe_stdin(char  **argv);
@@ -53,7 +58,7 @@ int     sig;
     exit(0);
 }
 
-main(argc, argv)
+int main(argc, argv)
 int     argc;
 char  **argv;
 {
diff --git a/tcpd.c b/tcpd.c
index 6c98966..d3f6e50 100644
--- a/tcpd.c
+++ b/tcpd.c
@@ -24,6 +24,7 @@ static char sccsid[] = "@(#) tcpd.c 1.10 96/02/11 17:01:32";
 #include <stdio.h>
 #include <syslog.h>
 #include <string.h>
+#include <unistd.h>
 
 #ifndef MAXPATHNAMELEN
 #define MAXPATHNAMELEN	BUFSIZ
diff --git a/tcpdchk.c b/tcpdchk.c
index 64665e1..81d1318 100644
--- a/tcpdchk.c
+++ b/tcpdchk.c
@@ -30,6 +30,7 @@ static char sccsid[] = "@(#) tcpdchk.c 1.8 97/02/12 02:13:25";
 #include <errno.h>
 #include <netdb.h>
 #include <string.h>
+#include <unistd.h>
 
 extern int errno;
 extern void exit();
diff --git a/tcpdmatch.c b/tcpdmatch.c
index 1142c93..776972c 100644
--- a/tcpdmatch.c
+++ b/tcpdmatch.c
@@ -29,6 +29,8 @@ static char sccsid[] = "@(#) tcpdmatch.c 1.5 96/02/11 17:01:36";
 #include <syslog.h>
 #include <setjmp.h>
 #include <string.h>
+#include <stdlib.h>
+#include <unistd.h>
 
 extern void exit();
 extern int optind;
diff --git a/try-from.c b/try-from.c
index 925e144..a246d4c 100644
--- a/try-from.c
+++ b/try-from.c
@@ -37,7 +37,7 @@ static char sccsid[] = "@(#) try-from.c 1.2 94/12/28 17:42:55";
 int     allow_severity = SEVERITY;	/* run-time adjustable */
 int     deny_severity = LOG_WARNING;	/* ditto */
 
-main(argc, argv)
+int main(argc, argv)
 int     argc;
 char  **argv;
 {
diff --git a/update.c b/update.c
index 6cf758b..52299e9 100644
--- a/update.c
+++ b/update.c
@@ -20,6 +20,7 @@ static char sccsid[] = "@(#) update.c 1.1 94/12/28 17:42:56";
 /* System libraries */
 
 #include <stdio.h>
+#include <unistd.h>
 #include <syslog.h>
 #include <string.h>
 
